cps_data/pycps/finalprep.py

```diff
@@ -0,0 +1,285 @@
+    goal_total = (var * weight).sum()
+    # Goal distribution based on IRS data
+    distribution = target / target.sum()
+    # Find the goal amount in each bin
+    goal_amts = goal_total * distribution
+    assert np.allclose(goal_amts.sum(), goal_total)
+    # Find current totals in each bin
+    bin_0 = np.where(agi < 0,
+                     var * weight, 0).sum()
+    bin_1 = np.where((agi >= 0) & (agi < 5000),
+                     var * weight, 0).sum()
+    bin_2 = np.where((agi >= 5000) & (agi < 10000),
+                     var * weight, 0).sum()
+    bin_3 = np.where((agi >= 10000) & (agi < 15000),
+                     var * weight, 0).sum()
+    bin_4 = np.where((agi >= 15000) & (agi < 20000),
+                     var * weight, 0).sum()
+    bin_5 = np.where((agi >= 20000) & (agi < 25000),
+                     var * weight, 0).sum()
+    bin_6 = np.where((agi >= 25000) & (agi < 30000),
+                     var * weight, 0).sum()
+    bin_7 = np.where((agi >= 30000) & (agi < 40000),
+                     var * weight, 0).sum()
+    bin_8 = np.where((agi >= 40000) & (agi < 50000),
+                     var * weight, 0).sum()
+    bin_9 = np.where((agi >= 50000) & (agi < 75000),
+                     var * weight, 0).sum()
+    bin_10 = np.where((agi >= 75000) & (agi < 100000),
+                      var * weight, 0).sum()
+    bin_11 = np.where((agi >= 100000) & (agi < 200000),
+                      var * weight, 0).sum()
+    bin_12 = np.where((agi >= 200000) & (agi < 500000),
+                      var * weight, 0).sum()
+    bin_13 = np.where((agi >= 500000) & (agi < 1e6),
+                      var * weight, 0).sum()
+    bin_14 = np.where((agi >= 1e6) & (agi < 1.5e6),
+                      var * weight, 0).sum()
+    bin_15 = np.where((agi >= 1.5e6) & (agi < 2e6),
+                      var * weight, 0).sum()
+    bin_16 = np.where((agi >= 2e6),
+                      var * weight, 0).sum()
+    # Create series holding each of the current totals
+    actual_amts = pd.Series([bin_0, bin_1, bin_2, bin_3, bin_4, bin_5,
+                             bin_6, bin_7, bin_8, bin_9, bin_10, bin_11,
+                             bin_12, bin_13, bin_14, bin_15, bin_16],
+                            index=goal_amts.index)
+    ratios_index = [num for num in range(0, len(actual_amts))]
+    # Determine the ratios
+    ratios = pd.Series(goal_amts / actual_amts)
+    ratios.index = ratios_index
+
+    # Apply adjustment ratios
+    var_array = np.array(var)
+    var_array = np.nan_to_num(var_array)
+    ratios = np.where(ratios == np.inf, 1., ratios)
+    adj_array = ratios[agi_bin]
+    var *= adj_array
+    # import pdb; pdb.set_trace()
+
+    # assert that we don't lose any of the variable
+    tot = (var*weight).sum()
+    m = f'{tot}:::{goal_total}'
+
+    assert np.allclose(np.array(goal_total), np.array(tot)), m
+
+    return var
+
+
+def adjust(data, targets):
```

Comment:
@andersonfrailey, it doesn't appear that `adjust()` or `adjust_helper()` is taking into account the filer/non-filer distinction when creating the adjustment factors. 

`target()` seems very similar to `adjust,` just focussed on a wider variety of state-based adjustments. (@andersonfrailey, if I am missing anything in that assumption I'd love a correction). 

`target()` calculates the adjustment factor as the difference between the filer population's sum and the SOI target, and then it applies the adjustment factor across both filers and non-filers. This generally makes sense to me (and seems like a fine approach for this PR since it doesn't seem to be a departure from the SAS code), but I want to think more about whether it ends up skewing the filer/non-filer count one way or the other.

Reply:
@MattHJensen, in addition to focusing on a wider variety of state-based adjustments, `target()` actually changes the total value of the targeted variables while `adjust()` only changes the distribution of the targeted variable in the data. We don't make a distinction between filers and non-filers in `adjust()` because we aren't changing the totals.

Neither of these would change the unweighted total of filers/non-filers because we make that distinction before any adjustments are made, but they could affect weighted sums after the stage 2 adjustments.

Reply:
Thanks Anderson. Resolving.